Library/Homebrew/os/linux/ld.rb

```diff
@@ -7,13 +7,36 @@ module Linux
     module Ld
       sig { returns(String) }
       def self.brewed_ld_so_diagnostics
+        @brewed_ld_so_diagnostics ||= T.let(nil, T.nilable(String))
+        @retried_brewed_ld_so_diagnostics ||= T.let(false, T.nilable(T::Boolean))
+
+        # Check for `#nil?` here because @brewed_ld_so_diagnostics could be a blank string.
+        return @brewed_ld_so_diagnostics unless @brewed_ld_so_diagnostics.nil?
+
         brewed_ld_so = HOMEBREW_PREFIX/"lib/ld.so"
         return "" unless brewed_ld_so.exist?
 
         ld_so_output = Utils.popen_read(brewed_ld_so, "--list-diagnostics")
-        return "" unless $CHILD_STATUS.success?
+        @brewed_ld_so_diagnostics = if $CHILD_STATUS.success?
```

Comment:
Ok reading this again has reminded me now why this wasn't cached. When we run `brew install hello` on a new system with an old glibc: the install cycle (ignoring `gcc` stuff) will be roughly: `linux-headers@5.15` then `glibc` then `hello`.

The first formula will set `ld.so` to system glibc and the second install will set `ld.so` to brewed `glibc`. So we may cache based on system glibc when it has changed under us to be a different glibc, which could result in failed linkage checks for the third formula if the system glibc is too old.

Reply:
But we can cache it if system `glibc` is new enough, right?

Reply:
Probably. Though `readlink` should be cheaper than checking that.

Reply:
Didn't see your latest edit -- could you sketch roughly what the `readlink` check should look like?

Reply:
Cached with a key based on the output of `#readlink` instead.